### DISCUSSION TITLE
WPF sample?

### DISCUSSION THREAD

**Discussion by jordi-z80** [kniEngine/kni#1794] [Q&A] [answered]
Hi, is there a working sample using KNI embedded in WPF (net core)?

I'm trying to use something like https://github.com/damian-666/monogame-framework-wpfinterop  It uses reflection in D3D11Host.cs to access some MonoGame internals (_d3dContext, ...) but as on KNI those variables are on different places, it just crashes. 

I've located the _d3dContext and I've been able to get some internals through reflection (strategy, context) to get to _d3dContext, but I get lost in the middle because I'm not familiar with KNI's structure.

Thanks!

**Comment by jordi-z80** [comment]
stupid me... I just found that you don't need to juggle as with MG to get the device context :)

This does the trick for the d3d context

```
private static DeviceContext GetSharpDxDeviceContextWithReflection(GraphicsDevice graphicsDevice)
{
    D3D11.Device dev = (D3D11.Device)graphicsDevice.GetD3D11Device ();
    DeviceContext ctx = dev.ImmediateContext;
    return ctx;
}
```

And this for the texture

```
private static Resource GetSharpDXResourceWithReflection(Texture renderTarget2D)
{
    return (Resource)renderTarget2D.GetD3D11Resource ();
}
```


There's another function for a multisampled texture but I can pass without it for the moment :)

**Reply by nkast** [reply, marked as answer]
Hello @jordi-z80 

I see you found the DX extensions.

FYI, I have a library for embedding KNI in WPF apps.
Here is a the sample https://github.com/nkast/ProtonType.Aether/tree/main/Samples/WpfInterop
The library is in the same repo.

**Comment by jordi-z80** [comment]
Thanks, the version I'm using has some annoying flickering when resizing that I'm not able to find, I'll try the library as soon as I can :)
